File: backend/app/schemas/recommendation.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class WordClickEvent(BaseModel):
    word: str
    context: str
    section: str = "Other"
    timestamp: float


class CopyEvent(BaseModel):
    text: str
    context: Optional[str] = None
    target_type: str = "text"  # "text", "translation", "summary"
    timestamp: float


class RecommendationSyncRequest(BaseModel):
    session_id: str
    paper_id: Optional[str] = None
    paper_title: Optional[str] = None
    paper_abstract: Optional[str] = None
    paper_keywords: Optional[List[str]] = None
    paper_difficulty: Optional[str] = None
    conversation_history: Optional[str] = None
    word_clicks: Optional[List[WordClickEvent]] = None
    copy_events: Optional[List[CopyEvent]] = None
    session_duration: Optional[float] = None
    is_final: bool = False


class RecommendationRolloutRequest(BaseModel):
    session_id: str
    user_score: int = Field(..., ge=1, le=10)
    user_comment: Optional[str] = None
    clicked_paper: Optional[str] = None
    followed_up_query: Optional[bool] = None


class RecommendationGenerateRequest(BaseModel):
    session_id: str
    user_query: Optional[str] = None  # ユーザーが希望する論文の種類（任意）


class RecommendationGenerateResponse(BaseModel):
    recommendations: List[Dict[str, Any]]
    reasoning: str
    knowledge_level: str
    search_queries: List[str]
    trace_id: Optional[str] = None

File: frontend/src/lib/recommendation.ts
import { API_URL } from "@/config";

export interface WordClickEvent {
	word: string;
	context: string;
	section: string;
	timestamp: number;
}

export interface RecommendationSyncPayload {
	session_id: string;
	paper_id?: string;
	paper_title?: string;
	paper_abstract?: string;
	paper_keywords?: string[];
	paper_difficulty?: string;
	conversation_history?: string;
	word_clicks?: WordClickEvent[];
	session_duration?: number;
}

export interface RecommendationFeedbackPayload {
	session_id: string;
	user_score: number;
	user_comment?: string;
	clicked_paper?: string;
	followed_up_query?: boolean;
}

export interface RecommendedPaper {
	title: string;
	abstract: string;
	url?: string;
	authors?: { name: string }[];
	year?: number;
	openAccessPdf?: { url: string };
	citationCount?: number;
}

export interface RecommendationGenerateResponse {
	recommendations: RecommendedPaper[];
	reasoning: string;
	knowledge_level: string;
	search_queries: string[];
}

/**
 * Build common headers for recommendation API requests.
 */
function buildHeaders(token: string | null): Record<string, string> {
	const headers: Record<string, string> = {
		"Content-Type": "application/json",
	};
	if (token) {
		headers.Authorization = `Bearer ${token}`;
	}
	return headers;
}

/**
 * Synchronize trajectory data with backend.
 * Fire-and-forget but errors are logged with details.
 */
export async function syncTrajectory(
	payload: RecommendationSyncPayload,
	token: string | null,
) {
	try {
		fetch(`${API_URL}/api/recommendation/sync`, {
			method: "POST",
			headers: buildHeaders(token),
			body: JSON.stringify(payload),
			keepalive: true,
		}).catch((err) => console.error("Failed to sync trajectory:", err));
	} catch (error) {
		console.error("Sync error:", error);
	}
}

/**
 * Submit feedback for recommendations.
 */
export async function submitRecommendationFeedback(
	payload: RecommendationFeedbackPayload,
	token: string | null,
) {
	const res = await fetch(`${API_URL}/api/recommendation/feedback`, {
		method: "POST",
		headers: buildHeaders(token),
		body: JSON.stringify(payload),
	});
	if (!res.ok) {
		const detail = await res.text().catch(() => "");
		throw new Error(`Failed to submit feedback (${res.status}): ${detail}`);
	}
	return res.json();
}

/**
 * Generate personal paper recommendations.
 */
export async function generateRecommendations(
	sessionId: string,
	token: string | null,
): Promise<RecommendationGenerateResponse> {
	const res = await fetch(`${API_URL}/api/recommendation/generate`, {
		method: "POST",
		headers: buildHeaders(token),
		body: JSON.stringify({ session_id: sessionId }),
	});
	if (!res.ok) {
		const detail = await res.text().catch(() => "");
		throw new Error(
			`Failed to generate recommendations (${res.status}): ${detail}`,
		);
	}
	return res.json();
}
